// @copilotkit/shared/dist/utils/errors.mjs
import {
  COPILOT_CLOUD_ERROR_NAMES,
  ConfigurationError,
  CopilotKitAgentDiscoveryError,
  CopilotKitApiDiscoveryError,
  CopilotKitError,
  CopilotKitErrorCode,
  CopilotKitLowLevelError,
  CopilotKitMisuseError,
  CopilotKitRemoteEndpointDiscoveryError,
  CopilotKitVersionMismatchError,
  ERROR_CONFIG,
  ERROR_NAMES,
  MissingPublicApiKeyError,
  ResolvedCopilotKitError,
  Severity,
  UpgradeRequiredError,
  getPossibleVersionMismatch
} from "../chunk-WIVBOP6V.mjs";
import "../chunk-VNNKZIFB.mjs";
import "../chunk-BANDZXMP.mjs";
import "../chunk-MSUB6DGR.mjs";
import "../chunk-IAFBVORQ.mjs";
import "../chunk-FCCOSO5L.mjs";
import "../chunk-PL5WNHFZ.mjs";
import "../chunk-JP2M4U4G.mjs";
import "../chunk-GYZIHHE6.mjs";
import "../chunk-P7STFMPO.mjs";
import "../chunk-LGWATNA5.mjs";
import "../chunk-6QGXWNS5.mjs";
import "../chunk-NAFEBKSO.mjs";
export {
  COPILOT_CLOUD_ERROR_NAMES,
  ConfigurationError,
  CopilotKitAgentDiscoveryError,
  CopilotKitApiDiscoveryError,
  CopilotKitError,
  CopilotKitErrorCode,
  CopilotKitLowLevelError,
  CopilotKitMisuseError,
  CopilotKitRemoteEndpointDiscoveryError,
  CopilotKitVersionMismatchError,
  ERROR_CONFIG,
  ERROR_NAMES,
  MissingPublicApiKeyError,
  ResolvedCopilotKitError,
  Severity,
  UpgradeRequiredError,
  getPossibleVersionMismatch
};
