// @copilotkit/react-ui/dist/chunk-Y7UO3RPW.mjs
import {
  Window
} from "./chunk-Q2NFQTCQ.mjs";
import {
  Header
} from "./chunk-UH2UFL5W.mjs";
import {
  Button
} from "./chunk-UFN2VWSR.mjs";
import {
  CopilotChat
} from "./chunk-O34Z4XM2.mjs";
import {
  UserMessage
} from "./chunk-HWMFMBJC.mjs";
import {
  AssistantMessage
} from "./chunk-OZXUB3V7.mjs";
import {
  Input
} from "./chunk-BDNHZ3GW.mjs";
import {
  Messages
} from "./chunk-IJADIQAR.mjs";
import {
  ChatContextProvider
} from "./chunk-IEMQ2SQW.mjs";

// src/components/chat/Modal.tsx
import React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var CopilotModal = ({
  instructions,
  defaultOpen = false,
  clickOutsideToClose = true,
  hitEscapeToClose = true,
  onSetOpen,
  onSubmitMessage,
  onStopGeneration,
  onReloadMessages,
  shortcut = "/",
  icons,
  labels,
  makeSystemMessage,
  onInProgress,
  Window: Window2 = Window,
  Button: Button2 = Button,
  Header: Header2 = Header,
  Messages: Messages2 = Messages,
  Input: Input2 = Input,
  AssistantMessage: AssistantMessage2 = AssistantMessage,
  UserMessage: UserMessage2 = UserMessage,
  onThumbsUp,
  onThumbsDown,
  onCopy,
  onRegenerate,
  className,
  children
}) => {
  const [openState, setOpenState] = React.useState(defaultOpen);
  const setOpen = (open) => {
    onSetOpen == null ? void 0 : onSetOpen(open);
    setOpenState(open);
  };
  return /* @__PURE__ */ jsxs(ChatContextProvider, { icons, labels, open: openState, setOpen, children: [
    children,
    /* @__PURE__ */ jsxs("div", { className, children: [
      /* @__PURE__ */ jsx(Button2, {}),
      /* @__PURE__ */ jsxs(
        Window2,
        {
          clickOutsideToClose,
          shortcut,
          hitEscapeToClose,
          children: [
            /* @__PURE__ */ jsx(Header2, {}),
            /* @__PURE__ */ jsx(
              CopilotChat,
              {
                instructions,
                onSubmitMessage,
                onStopGeneration,
                onReloadMessages,
                makeSystemMessage,
                onInProgress,
                Messages: Messages2,
                Input: Input2,
                AssistantMessage: AssistantMessage2,
                UserMessage: UserMessage2,
                onThumbsUp,
                onThumbsDown,
                onCopy,
                onRegenerate
              }
            )
          ]
        }
      )
    ] })
  ] });
};

export {
  CopilotModal
};
