// @h6s/calendar/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  CalendarViewType: () => CalendarViewType_default,
  arrayOf: () => arrayOf,
  generateID: () => generateID,
  isSameDate: () => isSameDate,
  isSameYearAndMonth: () => isSameYearAndMonth,
  parseDate: () => parseDate,
  pipe: () => pipe,
  pipeWith: () => pipeWith,
  useCalendar: () => useCalendar,
  withDateProps: () => withDateProps,
  withKey: () => withKey,
  withKeyProps: () => withKeyProps
});
module.exports = __toCommonJS(src_exports);

// src/models/CalendarViewType.ts
var CalendarViewtype = /* @__PURE__ */ ((CalendarViewtype2) => {
  CalendarViewtype2["Month"] = "month";
  CalendarViewtype2["Week"] = "week";
  CalendarViewtype2["Day"] = "day";
  return CalendarViewtype2;
})(CalendarViewtype || {});
var CalendarViewType_default = CalendarViewtype;

// src/plugins/withDateProps.ts
var import_date_fns4 = require("date-fns");

// src/utils/arrayOf.ts
function arrayOf(count) {
  return [...new Array(count).keys()];
}

// src/utils/generateID.ts
var randomId = 0;
var map = /* @__PURE__ */ new Map();
function generateID(prefix) {
  if (map.has(prefix)) {
    const id = map.get(prefix);
    const newId = id + 1;
    map.set(prefix, newId);
    randomId = newId;
  } else {
    const id = 1;
    map.set(prefix, id);
    randomId = id;
  }
  return `${prefix}-${randomId}`;
}

// src/utils/isSameDate.ts
var import_date_fns2 = require("date-fns");

// src/utils/resetTimeOfDate.ts
var import_date_fns = require("date-fns");
function resetTimeOfDate(date) {
  return (0, import_date_fns.setHours)((0, import_date_fns.setMinutes)((0, import_date_fns.setSeconds)((0, import_date_fns.setMilliseconds)(date, 0), 0), 0), 0);
}

// src/utils/isSameDate.ts
function isSameDate(baseDate, targetDate) {
  const base = resetTimeOfDate(baseDate);
  const target = resetTimeOfDate(targetDate);
  return (0, import_date_fns2.isEqual)(base, target);
}

// src/utils/isSameYearAndMonth.ts
var import_date_fns3 = require("date-fns");
function isSameYearAndMonth(baseDate, targetDate) {
  return (0, import_date_fns3.isSameMonth)(targetDate, baseDate) && (0, import_date_fns3.isSameYear)(targetDate, baseDate);
}

// src/utils/parseDate.ts
function parseDate(date) {
  return {
    year: date.getFullYear(),
    month: date.getMonth(),
    day: date.getDate()
  };
}

// src/utils/pipe.ts
function pipe(...fns) {
  return (...initialParams) => fns.reduce((value, fn, index) => {
    const params = index === 0 ? value : [value];
    return fn(...params);
  }, initialParams);
}
function pipeWith(value, ...fns) {
  return pipe(
    ...fns
  )(value);
}

// src/utils/withKey.ts
function withKey(arr, keyPrefix) {
  return arr.map((value) => {
    return {
      ...value,
      key: generateID(keyPrefix)
    };
  });
}

// src/plugins/withDateProps.ts
function withDateProps(baseDate, cursorDate) {
  return function(cell) {
    const { value: targetDate } = cell;
    const isCurrentMonth = isSameYearAndMonth(cursorDate, targetDate);
    const isCurrentDate = isSameDate(baseDate, targetDate);
    const isWeekend = (0, import_date_fns4.isWeekend)(targetDate);
    return {
      ...cell,
      date: (0, import_date_fns4.getDate)(targetDate),
      isCurrentMonth,
      isCurrentDate,
      isWeekend
    };
  };
}

// src/plugins/withKeyProps.ts
function withKeyProps(keyPrefix) {
  return function(cell) {
    return {
      ...cell,
      key: generateID(keyPrefix)
    };
  };
}

// src/useCalendar.ts
var import_date_fns6 = require("date-fns");
var import_react2 = require("react");

// src/core/createCalendarInfo.ts
var import_date_fns5 = require("date-fns");
function createCalendarInfo(cursorDate, { weekStartsOn }) {
  const { year, month, day } = parseDate(cursorDate);
  const startWeekdayInMonth = getStartWeekdayInMonth(cursorDate, weekStartsOn);
  const weeksInMonth = getWeeksInMonth(cursorDate, startWeekdayInMonth);
  const weekendDays = arrayOf(7).map((index) => ({
    value: (0, import_date_fns5.setDay)(cursorDate, index + weekStartsOn)
  }));
  const getDateCellByIndex = (weekIndex, dayIndex) => {
    const day2 = weekIndex * 7 + dayIndex - startWeekdayInMonth + 1;
    return { value: new Date(year, month, day2) };
  };
  return {
    cursorDate,
    year,
    month,
    day,
    weekStartsOn,
    startWeekdayInMonth,
    weeksInMonth,
    weekendDays,
    today: {
      weekIndex: getCurrentWeekIndex(day, startWeekdayInMonth),
      dateIndex: (0, import_date_fns5.getDay)(cursorDate)
    },
    getDateCellByIndex
  };
}
function getStartWeekdayInMonth(date, weekStartsOn) {
  const monthStartsAt = ((0, import_date_fns5.startOfMonth)(date).getDay() - weekStartsOn) % 7;
  return monthStartsAt < 0 ? monthStartsAt + 7 : monthStartsAt;
}
function getWeeksInMonth(date, startWeekdayInMonth) {
  const totalDaysOfMonth = (0, import_date_fns5.getDaysInMonth)(date);
  return Math.ceil((startWeekdayInMonth + totalDaysOfMonth) / 7);
}
function getCurrentWeekIndex(day, startWeekdayInMonth) {
  if ((day + startWeekdayInMonth) % 7 > 0) {
    return Math.floor((day + startWeekdayInMonth) / 7);
  }
  return Math.floor((day + startWeekdayInMonth) / 7) - 1;
}

// src/hooks/useIsMounted.tsx
var import_react = require("react");
function useIsMounted() {
  const [mounted, setMounted] = (0, import_react.useState)(false);
  (0, import_react.useEffect)(() => {
    setMounted(true);
  }, []);
  return mounted;
}

// src/useCalendar.ts
function useCalendar({
  defaultDate,
  defaultWeekStart = 0,
  defaultViewType = CalendarViewType_default.Month
} = {}) {
  const isMounted = useIsMounted();
  const baseDate = (0, import_react2.useMemo)(() => {
    return defaultDate != null ? new Date(defaultDate) : /* @__PURE__ */ new Date();
  }, [defaultDate, isMounted]);
  const [weekStartsOn, setWeekStartsOn] = (0, import_react2.useState)(defaultWeekStart);
  const [cursorDate, setCursorDate] = (0, import_react2.useState)(baseDate);
  const [viewType, setViewType] = (0, import_react2.useState)(defaultViewType);
  const calendar = createCalendarInfo(cursorDate, { weekStartsOn });
  const { weekendDays, weeksInMonth, today, getDateCellByIndex } = calendar;
  const getHeaders = (0, import_react2.useCallback)(
    (viewType2) => {
      switch (viewType2) {
        case CalendarViewType_default.Month:
        case CalendarViewType_default.Week:
          return {
            weekDays: withKey(weekendDays, "weekdays")
          };
        case CalendarViewType_default.Day:
        default:
          return {
            weekDays: withKey([{ value: cursorDate }], "weekdays")
          };
      }
    },
    [cursorDate, weekendDays]
  );
  const createMatrix = (0, import_react2.useCallback)(
    (weeksInMonth2) => ({
      value: arrayOf(weeksInMonth2).map((weekIndex) => {
        return {
          key: generateID("weeks"),
          value: arrayOf(7).map((dayIndex) => {
            return pipeWith(
              getDateCellByIndex(weekIndex, dayIndex),
              withDateProps(baseDate, cursorDate),
              withKeyProps("days")
            );
          })
        };
      })
    }),
    [baseDate, cursorDate, getDateCellByIndex]
  );
  const getBody = (0, import_react2.useCallback)(
    (viewType2) => {
      const matrix = createMatrix(weeksInMonth);
      const { weekIndex, dateIndex } = today;
      return {
        [CalendarViewType_default.Month]: matrix,
        [CalendarViewType_default.Week]: {
          value: [matrix.value[weekIndex]]
        },
        [CalendarViewType_default.Day]: {
          value: [
            {
              key: "week-day-type",
              value: [matrix.value[weekIndex]?.value[dateIndex]]
            }
          ]
        }
      }[viewType2];
    },
    [createMatrix, today, weeksInMonth]
  );
  const setNext = (0, import_react2.useMemo)(() => {
    switch (viewType) {
      case CalendarViewType_default.Month:
        return (date) => (0, import_date_fns6.addMonths)((0, import_date_fns6.startOfMonth)(date), 1);
      case CalendarViewType_default.Week:
        return (date) => (0, import_date_fns6.addWeeks)((0, import_date_fns6.startOfWeek)(date, { weekStartsOn }), 1);
      case CalendarViewType_default.Day:
        return (date) => (0, import_date_fns6.addDays)(date, 1);
    }
  }, [viewType, weekStartsOn]);
  const setPrev = (0, import_react2.useMemo)(() => {
    switch (viewType) {
      case CalendarViewType_default.Month:
        return (date) => (0, import_date_fns6.subMonths)((0, import_date_fns6.startOfMonth)(date), 1);
      case CalendarViewType_default.Week:
        return (date) => (0, import_date_fns6.subWeeks)((0, import_date_fns6.startOfWeek)(date, { weekStartsOn }), 1);
      case CalendarViewType_default.Day:
        return (date) => (0, import_date_fns6.subDays)(date, 1);
    }
  }, [viewType, weekStartsOn]);
  return (0, import_react2.useMemo)(
    () => ({
      ...calendar,
      headers: getHeaders(viewType),
      body: getBody(viewType),
      navigation: {
        toNext: () => setCursorDate((date) => setNext(date)),
        toPrev: () => setCursorDate((date) => setPrev(date)),
        setToday: () => setCursorDate(/* @__PURE__ */ new Date()),
        setDate: (date) => setCursorDate(date)
      },
      view: {
        type: viewType,
        setViewType,
        setWeekStartsOn,
        isMonthView: viewType === CalendarViewType_default.Month,
        isWeekView: viewType === CalendarViewType_default.Week,
        isDayView: viewType === CalendarViewType_default.Day,
        showMonthView: () => setViewType(CalendarViewType_default.Month),
        showWeekView: () => setViewType(CalendarViewType_default.Week),
        showDayView: () => setViewType(CalendarViewType_default.Day)
      }
    }),
    [calendar, getBody, getHeaders, setNext, setPrev, viewType]
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CalendarViewType,
  arrayOf,
  generateID,
  isSameDate,
  isSameYearAndMonth,
  parseDate,
  pipe,
  pipeWith,
  useCalendar,
  withDateProps,
  withKey,
  withKeyProps
});
