// @h6s/calendar/dist/index.js
// src/models/CalendarViewType.ts
var CalendarViewtype = /* @__PURE__ */ ((CalendarViewtype2) => {
  CalendarViewtype2["Month"] = "month";
  CalendarViewtype2["Week"] = "week";
  CalendarViewtype2["Day"] = "day";
  return CalendarViewtype2;
})(CalendarViewtype || {});
var CalendarViewType_default = CalendarViewtype;

// src/plugins/withDateProps.ts
import { getDate, isWeekend as _isWeekend } from "date-fns";

// src/utils/arrayOf.ts
function arrayOf(count) {
  return [...new Array(count).keys()];
}

// src/utils/generateID.ts
var randomId = 0;
var map = /* @__PURE__ */ new Map();
function generateID(prefix) {
  if (map.has(prefix)) {
    const id = map.get(prefix);
    const newId = id + 1;
    map.set(prefix, newId);
    randomId = newId;
  } else {
    const id = 1;
    map.set(prefix, id);
    randomId = id;
  }
  return `${prefix}-${randomId}`;
}

// src/utils/isSameDate.ts
import { isEqual } from "date-fns";

// src/utils/resetTimeOfDate.ts
import { setHours, setMilliseconds, setMinutes, setSeconds } from "date-fns";
function resetTimeOfDate(date) {
  return setHours(setMinutes(setSeconds(setMilliseconds(date, 0), 0), 0), 0);
}

// src/utils/isSameDate.ts
function isSameDate(baseDate, targetDate) {
  const base = resetTimeOfDate(baseDate);
  const target = resetTimeOfDate(targetDate);
  return isEqual(base, target);
}

// src/utils/isSameYearAndMonth.ts
import { isSameMonth as _isSameMonth, isSameYear } from "date-fns";
function isSameYearAndMonth(baseDate, targetDate) {
  return _isSameMonth(targetDate, baseDate) && isSameYear(targetDate, baseDate);
}

// src/utils/parseDate.ts
function parseDate(date) {
  return {
    year: date.getFullYear(),
    month: date.getMonth(),
    day: date.getDate()
  };
}

// src/utils/pipe.ts
function pipe(...fns) {
  return (...initialParams) => fns.reduce((value, fn, index) => {
    const params = index === 0 ? value : [value];
    return fn(...params);
  }, initialParams);
}
function pipeWith(value, ...fns) {
  return pipe(
    ...fns
  )(value);
}

// src/utils/withKey.ts
function withKey(arr, keyPrefix) {
  return arr.map((value) => {
    return {
      ...value,
      key: generateID(keyPrefix)
    };
  });
}

// src/plugins/withDateProps.ts
function withDateProps(baseDate, cursorDate) {
  return function(cell) {
    const { value: targetDate } = cell;
    const isCurrentMonth = isSameYearAndMonth(cursorDate, targetDate);
    const isCurrentDate = isSameDate(baseDate, targetDate);
    const isWeekend = _isWeekend(targetDate);
    return {
      ...cell,
      date: getDate(targetDate),
      isCurrentMonth,
      isCurrentDate,
      isWeekend
    };
  };
}

// src/plugins/withKeyProps.ts
function withKeyProps(keyPrefix) {
  return function(cell) {
    return {
      ...cell,
      key: generateID(keyPrefix)
    };
  };
}

// src/useCalendar.ts
import {
  addDays,
  addMonths,
  addWeeks,
  startOfMonth as startOfMonth2,
  startOfWeek,
  subDays,
  subMonths,
  subWeeks
} from "date-fns";
import { useCallback, useMemo, useState as useState2 } from "react";

// src/core/createCalendarInfo.ts
import { getDay, getDaysInMonth, setDay, startOfMonth } from "date-fns";
function createCalendarInfo(cursorDate, { weekStartsOn }) {
  const { year, month, day } = parseDate(cursorDate);
  const startWeekdayInMonth = getStartWeekdayInMonth(cursorDate, weekStartsOn);
  const weeksInMonth = getWeeksInMonth(cursorDate, startWeekdayInMonth);
  const weekendDays = arrayOf(7).map((index) => ({
    value: setDay(cursorDate, index + weekStartsOn)
  }));
  const getDateCellByIndex = (weekIndex, dayIndex) => {
    const day2 = weekIndex * 7 + dayIndex - startWeekdayInMonth + 1;
    return { value: new Date(year, month, day2) };
  };
  return {
    cursorDate,
    year,
    month,
    day,
    weekStartsOn,
    startWeekdayInMonth,
    weeksInMonth,
    weekendDays,
    today: {
      weekIndex: getCurrentWeekIndex(day, startWeekdayInMonth),
      dateIndex: getDay(cursorDate)
    },
    getDateCellByIndex
  };
}
function getStartWeekdayInMonth(date, weekStartsOn) {
  const monthStartsAt = (startOfMonth(date).getDay() - weekStartsOn) % 7;
  return monthStartsAt < 0 ? monthStartsAt + 7 : monthStartsAt;
}
function getWeeksInMonth(date, startWeekdayInMonth) {
  const totalDaysOfMonth = getDaysInMonth(date);
  return Math.ceil((startWeekdayInMonth + totalDaysOfMonth) / 7);
}
function getCurrentWeekIndex(day, startWeekdayInMonth) {
  if ((day + startWeekdayInMonth) % 7 > 0) {
    return Math.floor((day + startWeekdayInMonth) / 7);
  }
  return Math.floor((day + startWeekdayInMonth) / 7) - 1;
}

// src/hooks/useIsMounted.tsx
import { useEffect, useState } from "react";
function useIsMounted() {
  const [mounted, setMounted] = useState(false);
  useEffect(() => {
    setMounted(true);
  }, []);
  return mounted;
}

// src/useCalendar.ts
function useCalendar({
  defaultDate,
  defaultWeekStart = 0,
  defaultViewType = CalendarViewType_default.Month
} = {}) {
  const isMounted = useIsMounted();
  const baseDate = useMemo(() => {
    return defaultDate != null ? new Date(defaultDate) : /* @__PURE__ */ new Date();
  }, [defaultDate, isMounted]);
  const [weekStartsOn, setWeekStartsOn] = useState2(defaultWeekStart);
  const [cursorDate, setCursorDate] = useState2(baseDate);
  const [viewType, setViewType] = useState2(defaultViewType);
  const calendar = createCalendarInfo(cursorDate, { weekStartsOn });
  const { weekendDays, weeksInMonth, today, getDateCellByIndex } = calendar;
  const getHeaders = useCallback(
    (viewType2) => {
      switch (viewType2) {
        case CalendarViewType_default.Month:
        case CalendarViewType_default.Week:
          return {
            weekDays: withKey(weekendDays, "weekdays")
          };
        case CalendarViewType_default.Day:
        default:
          return {
            weekDays: withKey([{ value: cursorDate }], "weekdays")
          };
      }
    },
    [cursorDate, weekendDays]
  );
  const createMatrix = useCallback(
    (weeksInMonth2) => ({
      value: arrayOf(weeksInMonth2).map((weekIndex) => {
        return {
          key: generateID("weeks"),
          value: arrayOf(7).map((dayIndex) => {
            return pipeWith(
              getDateCellByIndex(weekIndex, dayIndex),
              withDateProps(baseDate, cursorDate),
              withKeyProps("days")
            );
          })
        };
      })
    }),
    [baseDate, cursorDate, getDateCellByIndex]
  );
  const getBody = useCallback(
    (viewType2) => {
      const matrix = createMatrix(weeksInMonth);
      const { weekIndex, dateIndex } = today;
      return {
        [CalendarViewType_default.Month]: matrix,
        [CalendarViewType_default.Week]: {
          value: [matrix.value[weekIndex]]
        },
        [CalendarViewType_default.Day]: {
          value: [
            {
              key: "week-day-type",
              value: [matrix.value[weekIndex]?.value[dateIndex]]
            }
          ]
        }
      }[viewType2];
    },
    [createMatrix, today, weeksInMonth]
  );
  const setNext = useMemo(() => {
    switch (viewType) {
      case CalendarViewType_default.Month:
        return (date) => addMonths(startOfMonth2(date), 1);
      case CalendarViewType_default.Week:
        return (date) => addWeeks(startOfWeek(date, { weekStartsOn }), 1);
      case CalendarViewType_default.Day:
        return (date) => addDays(date, 1);
    }
  }, [viewType, weekStartsOn]);
  const setPrev = useMemo(() => {
    switch (viewType) {
      case CalendarViewType_default.Month:
        return (date) => subMonths(startOfMonth2(date), 1);
      case CalendarViewType_default.Week:
        return (date) => subWeeks(startOfWeek(date, { weekStartsOn }), 1);
      case CalendarViewType_default.Day:
        return (date) => subDays(date, 1);
    }
  }, [viewType, weekStartsOn]);
  return useMemo(
    () => ({
      ...calendar,
      headers: getHeaders(viewType),
      body: getBody(viewType),
      navigation: {
        toNext: () => setCursorDate((date) => setNext(date)),
        toPrev: () => setCursorDate((date) => setPrev(date)),
        setToday: () => setCursorDate(/* @__PURE__ */ new Date()),
        setDate: (date) => setCursorDate(date)
      },
      view: {
        type: viewType,
        setViewType,
        setWeekStartsOn,
        isMonthView: viewType === CalendarViewType_default.Month,
        isWeekView: viewType === CalendarViewType_default.Week,
        isDayView: viewType === CalendarViewType_default.Day,
        showMonthView: () => setViewType(CalendarViewType_default.Month),
        showWeekView: () => setViewType(CalendarViewType_default.Week),
        showDayView: () => setViewType(CalendarViewType_default.Day)
      }
    }),
    [calendar, getBody, getHeaders, setNext, setPrev, viewType]
  );
}
export {
  CalendarViewType_default as CalendarViewType,
  arrayOf,
  generateID,
  isSameDate,
  isSameYearAndMonth,
  parseDate,
  pipe,
  pipeWith,
  useCalendar,
  withDateProps,
  withKey,
  withKeyProps
};
